target_test.go

```diff
@@ -0,0 +1,394 @@
+// The MIT License
+//
+// Copyright (c) 2020, Cloudflare, Inc. All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	odoh "github.com/cloudflare/odoh-go"
+	"github.com/miekg/dns"
+)
+
+type localResolver struct {
+	queries          []string
+	queryResponseMap map[string][]byte // Packed DNS queries to responses
+}
+
+func (r localResolver) name() string {
+	return "localResolver"
+}
+
+func (r localResolver) resolve(query *dns.Msg) (*dns.Msg, error) {
```

Comment:
Hmm, this is okay for most of the tests, but invalid resolution is not being tested. What is the result of an ODoH query that can be correctly decrypted, but the plaintext DNS query is invalid?

Reply:
We can do this in a future change.